packages/frontend/src/components/rich_text_editor/link_editor.tsx

```diff
@@ -0,0 +1,367 @@
+        }
+
+        const markAfter = childAfter.node.marks.find((mark) => mark.type === linkType);
+        if (!markAfter || markAfter.attrs.href !== href) {
+            break;
+        }
+
+        end = childAfter.offset + childAfter.node.nodeSize;
+    }
+
+    // Convert to absolute document positions
+    const linkStart = $pos.start() + start;
+    const linkEnd = $pos.start() + end;
+
+    const title = view.state.doc.textBetween(linkStart, linkEnd, "");
+    const { left, top } = getLinkCoords(view, linkStart);
+
+    return {
+        href,
+        title,
+        linkStart,
+        linkEnd,
+        left,
+        top,
+    };
+}
+
+function getLinkCoords(
+    view: EditorView,
+    linkPos: number,
+): {
+    left: number;
+    top: number;
+} {
+    const { node, offset } = view.domAtPos(linkPos);
+    const linkEl = node.childNodes[offset];
+
+    let linkRect: DOMRect;
+    if (linkEl && linkEl.nodeType === Node.ELEMENT_NODE) {
+        linkRect = (linkEl as Element).getBoundingClientRect();
+    } else {
+        console.error("no link element found");
+        // Defensive programming. I've never seen this branch used but there are no guarantees that
+        // `node.childNodes[offset]` is a link element. The types indicate that `node` may not be an
+        // HTML element, however that seems impossible when `linkStart` is in a link.
+        //
+        // Use the parent of the link as the basis for positioning the editor widget
+        linkRect = (node as Element).getBoundingClientRect();
+    }
+
+    const editorRect = view.dom.getBoundingClientRect();
+
+    const top = linkRect.bottom - editorRect.top + 5 + window.scrollY;
+    const left = linkRect.left - editorRect.left + window.scrollX;
+
+    return {
+        top,
+        left,
+    };
+}
+
+interface LinkEditorWidgetProps {
+    view: EditorView;
+    linkState: LinkState | null;
+    onChange: (attrs: { title: string; href: string }) => void;
+}
+const LinkEditorWidget: Component<LinkEditorWidgetProps> = (props) => {
+    function handleHrefInput(e: InputEvent) {
+        if (!props.linkState) {
+            return;
+        }
+
+        const newHref = (e.currentTarget as HTMLInputElement).value;
+        props.onChange({ href: newHref, title: props.linkState.title });
+    }
+
+    function handleTitleInput(e: InputEvent) {
+        if (!props.linkState) {
+            return;
+        }
+
+        const newTitle = (e.currentTarget as HTMLInputElement).value;
+        if (!newTitle) {
+            return;
+        }
+
+        props.onChange({ href: props.linkState.href, title: newTitle });
+    }
+
+    return (
+        <Show when={props.linkState} fallback={null}>
+            {(linkState) => (
+                <div
+                    class="link-editor-popup"
+                    style={{
+                        position: "absolute",
+                        top: `${linkState().top}px`,
+                        left: `${linkState().left}px`,
```

Comment:
(No action required.) It seems unfortunate to rely on manual coordinate calculations and absolute positioning, but a little searching suggests that this is recommended practice as the ProseMirror tooltip example works similarly: https://prosemirror.net/examples/tooltip/

Reply:
Yep, I agree that we're doing basically the same thing as the example.